packages/flutter/lib/src/cupertino/scaffold.dart

```diff
@@ -0,0 +1,212 @@
+    double topPadding = MediaQuery
+        .of(context)
+        .padding
+        .top;
+    if (widget.navigationBar is CupertinoNavigationBar) {
+      final CupertinoNavigationBar top = widget.navigationBar;
+      if (top.opaque)
+        topPadding += top.preferredSize.height;
+    }
+
+    double bottomPadding = 0.0;
+    if (widget.tabBar?.opaque ?? false)
+      bottomPadding = widget.tabBar.preferredSize.height;
+
+    return new Padding(
+      padding: new EdgeInsets.only(top: topPadding, bottom: bottomPadding),
+      child: middle,
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> stacked = <Widget>[];
+
+    // The main content being at the bottom is added to the stack first.
+    if (widget.child != null) {
+      stacked.add(_padMiddle(widget.child));
+    } else if (widget.rootTabPageBuilder != null) {
+      stacked.add(_padMiddle(new _TabView(
+        currentTabIndex: _currentPage,
+        tabNumber: widget.tabBar.items.length,
+        rootTabPageBuilder: widget.rootTabPageBuilder,
+      )));
+    }
+
+    if (widget.navigationBar != null) {
+      stacked.add(new Align(
+        alignment: FractionalOffset.topCenter,
+        child: widget.navigationBar,
+      ));
+    }
+
+    if (widget.tabBar != null) {
+      stacked.add(new Align(
+        alignment: FractionalOffset.bottomCenter,
+        child: widget.tabBar.copyWith(
+            currentIndex: _currentPage,
+            onTap: (int newIndex) {
+              setState(() {
+                _currentPage = newIndex;
+              });
+              // Chain the user's original callback after the automatic scaffold behavior.
+              if (widget.tabBar.onTap != null)
+                widget.tabBar.onTap(newIndex);
+            }
+        ),
+      ));
+    }
+
+    return new Stack(
+      children: stacked,
+    );
+  }
+}
+
+/// An widget laying out multiple tabs with only one active tab being built
+/// at a time and on stage. Off stage tabs' animations are stopped.
+class _TabView extends StatefulWidget {
+  _TabView({
+    @required this.currentTabIndex,
+    @required this.tabNumber,
+    @required this.rootTabPageBuilder,
+  }) : assert(currentTabIndex != null),
+       assert(tabNumber != null && tabNumber > 0),
+       assert(rootTabPageBuilder != null);
+
+  final int currentTabIndex;
+  final int tabNumber;
+  final IndexedWidgetBuilder rootTabPageBuilder;
+
+  @override
+  _TabViewState createState() => new _TabViewState();
+}
+
+class _TabViewState extends State<_TabView> {
+  List<Widget> tabs;
+
+  @override
+  void initState() {
+    super.initState();
+    tabs = new List<Widget>.filled(widget.tabNumber, new Container());
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new Stack(
+      children: new List<Widget>.generate(widget.tabNumber, (int index) {
+        final bool active = index == widget.currentTabIndex;
+        // If the tab is being shown re-build and cache it.
```

Comment:
These aren't edge cases. The contract when you have a builder is that you'll call it any time you build. Putting in old widgets is going to be a weird quirk that will lead to bugs. We should just call the builders.

It's fine to not call them until you need them, but once you're including them, you should call them (or not use a builder).